cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/task/ProvisioningTask.java

```diff
@@ -114,23 +119,33 @@ public void execute() throws Exception {
   }
 
   /**
-   * Write the task state to the {@link ProvisionerDataset}, retrying if any exception is caught.
+   * Write the task state to the {@link ProvisionerDataset}, retrying if any exception is caught. Before persisting
+   * the state, the current state will be checked. If the current state is cancelled, it will not be overwritten.
    *
    * @param taskInfo the task state to save
    * @param retryStrategy the retry strategy to use on errors
+   * @return the task info that is stored. This will be the taskInfo that was given to this method unless the existing
+   *   task info was in the cancelled state, in which case the cancelled info will be returned.
    * @throws TransactionFailureException if there was an error and the retry limit was hit
    * @throws InterruptedException if we were interrupted while waiting between retries
    */
-  protected void persistTaskInfo(ProvisioningTaskInfo taskInfo,
-                                 RetryStrategy retryStrategy) throws TransactionFailureException, InterruptedException {
+  protected ProvisioningTaskInfo persistTaskInfo(ProvisioningTaskInfo taskInfo, RetryStrategy retryStrategy)
+    throws TransactionFailureException, InterruptedException {
     try {
       // Stop retrying if we are interrupted. Otherwise, retry on every exception, up to the retry limit
-      Retries.callWithInterruptibleRetries(() -> {
+      return Retries.callWithInterruptibleRetries(() -> {
+        AtomicReference<ProvisioningTaskInfo> state = new AtomicReference<>(taskInfo);
```

Comment:
It'd be simpler to use `Transactionals.execute` and `TxCallable`, which the state is returned.

Reply:
That makes the exception handling more complicated though, as I would have to catch a RuntimeException and check the cause.

Reply:
I see. I don't know the exact exception since none of the dataset stuff throws checked exceptions. I think it's fine though, this class needs to handle any exception here so I can just change the method signature.